File: src/compatibility.py
import numpy as np
import pandas as pd
import logging

logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
logger = logging.getLogger(__name__)

def calculate_compatibility_index(row: pd.Series) -> float:
    """
    Calculates the Datezo Compatibility Index (0-100) using a transparent multi-attribute utility formula.
    
    Weights Breakdown:
    - Mutual Attractiveness: 20%
    - Shared Interests: 15%
    - Mutual Intelligence & Sincerity: 15%
    - Mutual Fun & Ambition: 15%
    - Preference Alignment: 10%
    - Perception Balance (Rating Gap): 10%
    - Background Compatibility (Race, Field, Age): 10%
    - Lifestyle Compatibility: 5%
    """
    # 1. Mutual Attraction (0-100)
    mutual_attr = float(row.get("mutual_attractiveness", (row.get("attr_of_female", 5) + row.get("attr_of_male", 5)) / 2.0))
    attr_score = np.clip((mutual_attr / 10.0) * 100.0, 0.0, 100.0)

    # 2. Shared Interests (0-100)
    shared_int = float(row.get("shared_interests", 5.0))
    interest_score = np.clip((shared_int / 10.0) * 100.0, 0.0, 100.0)

    # 3. Mutual Intelligence & Sincerity (0-100)
    mutual_intel = float(row.get("mutual_intelligence", (row.get("intel_of_female", 5) + row.get("intel_of_male", 5)) / 2.0))
    mutual_sinc = float(row.get("mutual_sincerity", (row.get("sinc_of_female", 5) + row.get("sinc_of_male", 5)) / 2.0))
    intel_sinc_score = np.clip(((mutual_intel + mutual_sinc) / 20.0) * 100.0, 0.0, 100.0)

    # 4. Mutual Fun & Ambition (0-100)
    mutual_fun = float(row.get("mutual_fun", (row.get("fun_of_female", 5) + row.get("fun_of_male", 5)) / 2.0))
    mutual_amb = float(row.get("mutual_ambition", (row.get("amb_of_female", 5) + row.get("amb_of_male", 5)) / 2.0))
    fun_amb_score = np.clip(((mutual_fun + mutual_amb) / 20.0) * 100.0, 0.0, 100.0)

    # 5. Preference Alignment (0-100)
    m_attr_align = float(row.get("male_attractiveness_alignment", 7.0))
    m_intel_align = float(row.get("male_intelligence_alignment", 7.0))
    f_attr_align = float(row.get("female_attractiveness_alignment", 7.0))
    f_intel_align = float(row.get("female_intelligence_alignment", 7.0))
    pref_score = np.clip(((m_attr_align + m_intel_align + f_attr_align + f_intel_align) / 40.0) * 100.0, 0.0, 100.0)

    # 6. Perception Balance (100 - rating gap penalty)
    rating_gap = float(row.get("overall_rating_gap", 1.0))
    balance_score = np.clip(100.0 - (rating_gap / 10.0 * 100.0), 0.0, 100.0)

    # 7. Background Compatibility
    same_race = float(row.get("same_race", 0))
    same_field = float(row.get("same_field", 0))
    age_comp = float(row.get("age_compatibility", 8.0))
    background_score = np.clip((same_race * 35.0 + same_field * 35.0 + (age_comp / 10.0) * 30.0), 0.0, 100.0)

    # 8. Lifestyle Compatibility
    life_comp = float(row.get("lifestyle_compatibility", 3.0))
    lifestyle_score = np.clip((life_comp / 4.0) * 100.0, 0.0, 100.0)

    # Weighted Sum
    compatibility_index = (
        0.20 * attr_score +
        0.15 * interest_score +
        0.15 * intel_sinc_score +
        0.15 * fun_amb_score +
        0.10 * pref_score +
        0.10 * balance_score +
        0.10 * background_score +
        0.05 * lifestyle_score
    )

    return round(float(compatibility_index), 1)

def get_compatibility_category(probability: float) -> str:
    """
    Categorizes the calibrated match probability into standard UI labels.
    """
    prob_percent = probability * 100.0 if probability <= 1.0 else probability
    if prob_percent < 30.0:
        return "Low Compatibility"
    elif prob_percent < 60.0:
        return "Moderate Compatibility"
    elif prob_percent < 80.0:
        return "High Compatibility"
    else:
        return "Very High Compatibility"

if __name__ == "__main__":
    test_row = pd.Series({
        "attr_of_female": 8.0, "attr_of_male": 8.0,
        "shared_interests": 8.5,
        "intel_of_female": 8.0, "intel_of_male": 8.0,
        "sinc_of_female": 8.0, "sinc_of_male": 8.0,
        "fun_of_female": 8.0, "fun_of_male": 8.0,
        "amb_of_female": 7.0, "amb_of_male": 7.0,
        "male_attractiveness_alignment": 8.5, "female_attractiveness_alignment": 8.5,
        "male_intelligence_alignment": 8.5, "female_intelligence_alignment": 8.5,
        "overall_rating_gap": 0.2,
        "same_race": 1, "same_field": 1, "age_compatibility": 9.0,
        "lifestyle_compatibility": 4
    })
    idx = calculate_compatibility_index(test_row)
    cat = get_compatibility_category(85.0)
    print(f"Test Compatibility Index: {idx}/100, Category: {cat}")

File: src/utils/compatibility.js
export function calculate_compatibility_index(data) {
  const mutualAttr = (Number(data.attr_of_female || 5) + Number(data.attr_of_male || 5)) / 2.0;
  const attrScore = Math.min(100, Math.max(0, (mutualAttr / 10.0) * 100.0));

  const sharedInt = Number(data.shared_interests || 5.0);
  const interestScore = Math.min(100, Math.max(0, (sharedInt / 10.0) * 100.0));

  const mutualIntel = (Number(data.intel_of_female || 5) + Number(data.intel_of_male || 5)) / 2.0;
  const mutualSinc = (Number(data.sinc_of_female || 5) + Number(data.sinc_of_male || 5)) / 2.0;
  const intelSincScore = Math.min(100, Math.max(0, ((mutualIntel + mutualSinc) / 20.0) * 100.0));

  const mutualFun = (Number(data.fun_of_female || 5) + Number(data.fun_of_male || 5)) / 2.0;
  const mutualAmb = (Number(data.amb_of_female || 5) + Number(data.amb_of_male || 5)) / 2.0;
  const funAmbScore = Math.min(100, Math.max(0, ((mutualFun + mutualAmb) / 20.0) * 100.0));

  const normMalePrefAttr = Number(data.male_pref_attr || 25) / 10.0;
  const normMalePrefIntel = Number(data.male_pref_intel || 25) / 10.0;
  const normFemalePrefAttr = Number(data.female_pref_attr || 25) / 10.0;
  const normFemalePrefIntel = Number(data.female_pref_intel || 25) / 10.0;

  const mAttrAlign = 10.0 - Math.abs(normMalePrefAttr - Number(data.attr_of_female || 5));
  const mIntelAlign = 10.0 - Math.abs(normMalePrefIntel - Number(data.intel_of_female || 5));
  const fAttrAlign = 10.0 - Math.abs(normFemalePrefAttr - Number(data.attr_of_male || 5));
  const fIntelAlign = 10.0 - Math.abs(normFemalePrefIntel - Number(data.intel_of_male || 5));

  const prefScore = Math.min(100, Math.max(0, ((mAttrAlign + mIntelAlign + fAttrAlign + fIntelAlign) / 40.0) * 100.0));

  const overallFemale = (Number(data.attr_of_female||5) + Number(data.sinc_of_female||5) + Number(data.intel_of_female||5) + Number(data.fun_of_female||5) + Number(data.amb_of_female||5)) / 5.0;
  const overallMale = (Number(data.attr_of_male||5) + Number(data.sinc_of_male||5) + Number(data.intel_of_male||5) + Number(data.fun_of_male||5) + Number(data.amb_of_male||5)) / 5.0;
  const ratingGap = Math.abs(overallFemale - overallMale);
  const balanceScore = Math.min(100, Math.max(0, 100.0 - (ratingGap / 10.0 * 100.0)));

  const sameRace = Number(data.same_race || 0);
  const sameField = Number(data.same_field || 0);
  const ageGap = Math.abs(Number(data.male_age || 25) - Number(data.female_age || 25));
  const ageComp = Math.min(10, Math.max(0, 10.0 - ageGap));
  const backgroundScore = Math.min(100, Math.max(0, (sameRace * 35.0 + sameField * 35.0 + (ageComp / 10.0) * 30.0)));

  const lifestyleMap = { rarely: 1, sometimes: 2, often: 3, very_often: 4 };
  const maleLife = lifestyleMap[String(data.male_goes_out).toLowerCase()] || 2;
  const femaleLife = lifestyleMap[String(data.female_goes_out).toLowerCase()] || 2;
  const lifeGap = Math.abs(maleLife - femaleLife);
  const lifestyleScore = Math.min(100, Math.max(0, ((4 - lifeGap) / 4.0) * 100.0));

  const compIndex = (
    0.20 * attrScore +
    0.15 * interestScore +
    0.15 * intelSincScore +
    0.15 * funAmbScore +
    0.10 * prefScore +
    0.10 * balanceScore +
    0.10 * backgroundScore +
    0.05 * lifestyleScore
  );

  return Math.round(compIndex * 10) / 10.0;
}

export function get_compatibility_category(probability) {
  const p = probability <= 1.0 ? probability * 100.0 : probability;
  if (p < 30.0) return "Low Compatibility";
  if (p < 60.0) return "Moderate Compatibility";
  if (p < 80.0) return "High Compatibility";
  return "Very High Compatibility";
}
